v-next/hardhat/src/internal/builtin-plugins/coverage/hook-handlers/solidity.ts

```diff
@@ -1,53 +1,91 @@
 import type { SolidityHooks } from "../../../../types/hooks.js";
+import type { CoverageMetadata } from "../types.js";
 
 import path from "node:path";
 
 import { addStatementCoverageInstrumentation } from "@ignored/edr-optimism";
 import { HardhatError } from "@nomicfoundation/hardhat-errors";
 import { ensureError } from "@nomicfoundation/hardhat-utils/error";
 import { readUtf8File } from "@nomicfoundation/hardhat-utils/fs";
+import debug from "debug";
 
 import { unsafelyCastAsHardhatRuntimeEnvironmentImplementation } from "../helpers.js";
 
+const log = debug("hardhat:core:coverage:hook-handlers:solidity");
+
 const COVERAGE_LIBRARY_PATH =
   "__hardhat_coverage_library_a3e9cfe2-41b4-4a1f-ad9e-ac62dd82979e.sol";
 
 export default async (): Promise<Partial<SolidityHooks>> => ({
   preprocessProjectFileBeforeBuilding: async (
     context,
     sourceName,
+    fsPath,
     fileContent,
     solcVersion,
     next,
   ) => {
-    if (context.globalOptions.coverage) {
+    // NOTE: We do not want to instrument the test project files as we don't
+    // want to report coverage for them.
+    const isTestSource =
+      fsPath.startsWith(context.config.paths.tests.solidity) ||
+      fsPath.endsWith(".t.sol");
```

Comment:
This is relevant when we execute the solidity tests (`hardhat test solidity` or `hardhat test` commands). Then, in addition to the sources, we also compile the test sources. We don't want to instrument the latter because the user should not care about the coverage of their test files. Instead, they care what percentage of their main sources their test sources cover.